Make GC bridge perform well in the "double fan" scenario

xmcclure opened the pull request:
The GC bridge currently performs VERY poorly when the object graph has a "double fan" shape. If M Java objects point to 1 C# object which points to N Java objects, the graph exported to Monodroid will currently drop the C# object node and replace it with (M\*N) edges. It is easy to write code where (M\*N) grows ludicrously big.

In testing on device, this patch solves that problem while leaving other cases unaffected. In testing on device, I found:
* All performance tests except double fan: No discernible performance difference after patch
* Double fan test with M=N=1000: Before the patch this took between three and seven seconds. After the patch this took between 0.2 and 0.3 seconds.
* Double fan test with M=N=4000: Before the patch this took anywhere from one to two minutes. After the patch, this took about 1.2 seconds.
* Double fan test with M=N=6000: Before the patch, this took so long I was not able to get it to ever complete. After the patch, this took about 1.3 seconds.
* Double fan test with M=N=20000: I did not attempt this test pre-patch. After the patch, it took about 5.6 seconds.

The commit messages describe the changes in some detail but the short version is:
* Add a mechanism for exporting non-bridged SCCs to the bridge client
* Turn that mechanism on whenever the product fanin*fanout grows too large
* Make the merge cache more aggressive

To function, this patch requires a corresponding change to monodroid. See https://github.com/xamarin/xamarin-android/pull/154

xmcclure commented on docs/sources/mono-api-gc.html:
```diff
@@ -40,9 +40,8 @@
 	`is_alive` flag on those strongly connected components that it
-	wishes to be kept alive.  Only bridged objects will be
-	reported to the callback, i.e., non-bridged objects are
-	removed from the callback graph.
```
Do we even still use this file?

kumpera replied:
This used to be generated AFAICT.

xmcclure commented on mono/metadata/sgen-tarjan-bridge.c:
```diff
@@ -124,7 +130,26 @@ typedef struct _ScanData {
 } ScanData;
 
+#define HEAVY_REFS_MIN 2
+#define HEAVY_COMBINED_REFS_MIN 60
```
This is slightly arbitrary and I do not have any complex Monodroid programs handy to fine-tune values against. I picked it such that if the incoming_colors ever hits its maximum size, the HEAVY_COMBINED_REFS_MIN would be sure to trigger. I note that if you hit the HEAVY_COMBINED_REFS_MIN then that means you are generating half a kilobyte of xrefs just for the list.

kumpera replied:
It would be nice to move those knobs and the one before (INCOMING_COLORS_BITS) together and include a description of how you picked the values and what's the impact in changing them.

Model it after this: https://github.com/mono/mono/blob/master/mono/sgen/sgen-conf.h

xmcclure commented on mono/metadata/sgen-tarjan-bridge.c:
```diff
@@ -398,6 +439,13 @@ dyn_array_ptr_contains (DynPtrArray *da, void *x)
+
+
+static gboolean
 match_colors (DynPtrArray *a, DynPtrArray *b)
```
Isn't this function a little wasteful anyway? It looks like it checks to test that A and B contain the same items regardless of order. However, we could get away with just checking that the arrays are equal. If the arrays contained the same items in a different order, we wouldn't even get here, because the hashes would be different.

kumpera replied:
That's a fair point.

kumpera replied:
Hashing is order independent. It does by adding the hashed version of each color pointer together.

Having said that. Adding elements when hashing is pretty bad as it has low avalanche properties, but it's the easy way around getting around the ordering problem.

The hash_perturb trick doesn't seem to help that much.

A simple test shows that the extra xoring introduces little to no impact into the actual hashcode, it cause some mild disturbance of low bits and none to the high bits. That together with the fact we do (hash & 0x7F) to compute the bucket translates into very little change. A simple test program shown it to a few times lead to streaks of repeating numbers. We're better off using a PRNG and stepping it on every collection.

This change will make debugging harder, as bugs repros won't be as stable.

The change to conservatively merge based on the hashcode makes a bit nervous, since we might end up merging too much and screwing up things. We should at least make it runtime toggleable in case we need to disable it for a problematic customer while we troubleshot.

xmcclure replied:
Ugh, I was not paying attention and misunderstood what the hash function was doing. Let me take a look at this...

xmcclure commented on mono/metadata/sgen-tarjan-bridge.c:
```diff
@@ -411,33 +459,43 @@ match_colors (DynPtrArray *a, DynPtrArray *b)
-			return bucket [i].color;
+
+		/* Color equality checking is very expensive with a lot of elements, so when there are many
+		 * elements we switch to a cheap comparison method which allows false positives. When false
```
Notice this

jonpryor commented on mono/metadata/sgen-tarjan-bridge.c:
```diff
@@ -94,13 +92,21 @@ enum {
 */
+#define API_INDEX_BITS        26
+#define INCOMING_COLORS_BITS  5
+#define API_INDEX_MAX         ((1<<API_INDEX_BITS)-1)
```
Does mono have a "static assert" macro? If so, we should assert that `(API_INDEX_BITS + INCOMING_COLORS_BITS) <= 31`.

xmcclure replied:
What I'll do is add a normal assert in `sgen_tarjan_bridge_init`. We already have some "static-ish" asserts there.

evincarofautumn replied:
In this case, the preprocessor can do it:

```
#if defined (API_INDEX_BITS) && defined (INCOMING_COLORS_BITS)
#   if !(API_INDEX_BITS + INCOMING_COLORS_BITS <= 31)
#       error "Assertion failed!"
#   endif
#else
#    error "Someone renamed these #defines and didn't update the assertion."
#endif
```

I don’t suppose we can use C11 `static_assert`.

jonpryor commented on mono/metadata/sgen-tarjan-bridge.c:
```diff
@@ -115,7 +121,7 @@ typedef struct _ScanData {
 	int index;
 	// Tarjan index of lowest-index object known reachable from here
-	int low_index : 27;
+	signed low_index : 27;
```
why `s/int/signed`? `int` is implicitly signed.

xmcclure replied:
Something horrible I learned while writing this patch: No, for bitfields, it isn't :( From the [C11 standard](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf), section 6.7.2 around page 111 specifies that "unsigned, or unsigned int" designate the same type and that "int, signed or signed int" also designate the same type:

> except that for bit-fields, it is implementation-defined whether the specifier int designates the same type as signed int or the same type as unsigned int.

This is reiterated several times in the document, and there is similar language in the C89 standard :(

It might be safe in practice but I'm considering going through and mass switching our `int` bitfields to explicitly signed.

evincarofautumn replied:
I looked into it. `signed` is the default for GCC, Clang, MSVC, and ICC, so it looks like we’re safe. We could add `-fsigned-bitfields` for GCC and Clang, but it’s the default; Clang doesn’t even support `-funsigned-bitfields`, actually.

I would be more concerned about using `gboolean flag : 1;` and having a true value get silently truncated to false.

xmcclure replied:
C11 actually has a type for boolean-semantics bitfields… if you're using C11…

jonpryor commented on mono/metadata/sgen-tarjan-bridge.c:
```diff
@@ -124,7 +130,26 @@ typedef struct _ScanData {
+ * Notice fanout is capped at INCOMING_COLORS_MAX so we can't know its exact true value.
+ */
+static inline gboolean bridgeless_color_is_heavy (ColorData *data) {
+   int fanin = data->incoming_colors;
```
Indentation is inconsistent within this patch. Here, you're using 2 space indents. In `color_merge_array_empty()`, you're using tabs.

I don't know if this file should be using tabs or spaces, but presumably the patch itself could be consistent?

xmcclure replied:
will fix

kumpera replied:
return type goes to a separate line as well.

it's ```c
static inline gboolean
bridgeless_color_is_heavy (ColorData *data)
{
...
```

kumpera commented on mono/metadata/sgen-tarjan-bridge.c:
```diff
@@ -124,7 +130,26 @@ typedef struct _ScanData {
+#define HEAVY_COMBINED_REFS_MIN 60
 
+/* Should color be made visible to client even though it has no bridges?
+ * If fanin and fanout are both greater than 2, then returning TRUE here will add one node to
```
not greater than 2, but greater than HEAVY_REFS_MIN

xmcclure commented:
Addressed above comments. Please see commit messages.

Re: the aggressive merging and some hash "perturb":

I did some testing. If you feed 50,000 random numbers into the hash function (this is roughly what we're doing during a GC in the worst case, since realistically there are no more than 45,000 bridged objects at any one time), you'll get a collision about 1/4 of the time (that is, in one out of every 4 worst-case GCs there will exist two bridge objects who produce the same hash). You'll get more than one collision about 1/40 of the time (that is, in about 1 out of every 40 worst-case GCs). So I think this is probably safe, but the thing that worries me is if two objects collide systematically (ie they get inappropriately merged on *every* GC). This is what the perturb addresses.

In the new commit, the perturb is injected before the hash occurs, which means that even though perturbing only affects a couple bits the effects of those bits are distributed throughout the final hash value. In my tests, if I took two values which hash-collide and applied the perturb before hashing, this was extremely effective in causing the values to no longer collide after. This was with the scheme where the perturb value only increments, and since the increment scheme works well in testing I would prefer it over a PRNG because less code is needed and we don't have to think about period.

kumpera commented on mono/metadata/sgen-bridge.c:
```diff
@@ -655,19 +668,24 @@ register_test_bridge_callbacks (const char *bridge_class_name)
 	} else if (!strcmp (opt, "enable-bridge-accounting")) {
-		bridge_processor.enable_accounting ();
+		bridge_processor_config.accounting = TRUE;
+	} else if (!strcmp (opt, "enable-tarjan-precise-merge")) {
```
This needs to go to man page and release notes.

xmcclure replied:
I'm filing a request with the docs team to get it documented in the appropriate place for Android.

kumpera commented:
LGTM.

xmcclure commented:
@monojenkins merge
